File: engine/dynamic/render_vulkan/src/renderer/shader_mgmt.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/logging.hpp"

#include "argus/resman.hpp"

#include "argus/render/defines.h"
#include "argus/render/common/shader_compilation.hpp"

#include "internal/render_vulkan/renderer/shader_mgmt.hpp"
#include "internal/render_vulkan/setup/device.hpp"

#include "glslang/Public/ShaderLang.h"

#include <iterator>
#include <ostream>
#include <sstream>
#include <string>
#include <vector>

namespace argus {
    struct ShaderCompilationResult {
        std::vector<Shader> shaders;
        ShaderReflectionInfo reflection;
    };

    static ShaderCompilationResult _compile_glsl_shaders(const std::vector<Shader> &shaders) {
        std::vector<std::pair<Shader, VkShaderModule>> handles;

        if (shaders.empty()) {
            return ShaderCompilationResult {};
        }

        std::vector<std::string> shader_uids;
        std::vector<std::string> shader_sources;
        for (auto &shader : shaders) {
            shader_uids.emplace_back(shader.get_uid());
            shader_sources.emplace_back(shader.get_source().begin(), shader.get_source().end());
        }

        std::ostringstream shader_uids_oss;
        std::copy(shader_uids.begin(), shader_uids.end(),
                std::ostream_iterator<std::string>(shader_uids_oss, ", "));
        auto shader_uids_str = shader_uids_oss.str().substr(0, shader_uids_oss.str().size() - strlen(", "));
        Logger::default_logger().debug("Compiling SPIR-V from shader set [%s]", shader_uids_str.c_str());

        auto [spirv_shaders, refl_info] = compile_glsl_to_spirv(shaders, glslang::EShClientVulkan,
                glslang::EShTargetVulkan_1_2, glslang::EShTargetSpv_1_0);

        ShaderCompilationResult res;
        res.reflection = refl_info;

        for (auto &shader : spirv_shaders) {
            Logger::default_logger().debug("Creating shader %s", shader.get_uid().c_str());
            res.shaders.emplace_back(shader);
        }

        return res;
    }

    PreparedShaderSet prepare_shaders(VkDevice device, std::initializer_list<std::string> shader_uids) {
        auto vec = std::vector<std::string>(shader_uids);
        return prepare_shaders(device, vec);
    }

    PreparedShaderSet prepare_shaders(VkDevice device, const std::vector<std::string> &shader_uids) {
        PreparedShaderSet res;

        std::vector<Resource *> shader_resources;
        std::vector<Shader> loaded_shaders;
        bool have_vert = false;
        bool have_frag = false;
        for (const auto &shader_uid : shader_uids) {
            auto &shader_res = ResourceManager::instance().get_resource(shader_uid)
                    .expect("Failed to load shader " + shader_uid);
            auto &shader = shader_res.get<Shader>();

            shader_resources.push_back(&shader_res);
            loaded_shaders.push_back(shader);

            if (shader.get_stage() == ShaderStage::Vertex) {
                have_vert = true;
            } else if (shader.get_stage() == ShaderStage::Fragment) {
                have_frag = true;
            }
        }

        if (!have_vert) {
            auto &vert_res = ResourceManager::instance().get_resource(SHADER_STD_VERT)
                    .expect("Failed to load built-in shader " SHADER_STD_VERT);
            shader_resources.push_back(vert_res);
            loaded_shaders.push_back(vert_res.get<Shader>());
        }
        if (!have_frag) {
            auto &frag_res = ResourceManager::instance().get_resource(SHADER_STD_FRAG)
                    .expect("Failed to load built-in shader " SHADER_STD_FRAG);
            shader_resources.push_back(frag_res);
            loaded_shaders.push_back(frag_res.get<Shader>());
        }

        //TODO: handle native SPIR-V shaders too
        auto comp_res = _compile_glsl_shaders(loaded_shaders);
        res.reflection = comp_res.reflection;

        for (const auto &shader : comp_res.shaders) {
            auto stage = shader.get_stage();
            auto spirv_src = shader.get_source();

            VkShaderStageFlagBits vk_shader_stage;
            switch (stage) {
                case ShaderStage::Vertex:
                    vk_shader_stage = VK_SHADER_STAGE_VERTEX_BIT;
                    break;
                case ShaderStage::Fragment:
                    vk_shader_stage = VK_SHADER_STAGE_FRAGMENT_BIT;
                    break;
                default:
                    crash("Unrecognized shader stage ordinal %d",
                            static_cast<std::underlying_type<ShaderStage>::type>(stage));
            }

            VkShaderModuleCreateInfo stage_create_info {};
            stage_create_info.sType = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO;
            stage_create_info.codeSize = shader.get_source().size();
            stage_create_info.pCode = reinterpret_cast<const uint32_t *>(shader.get_source().data());

            VkShaderModule shader_module;
            if (vkCreateShaderModule(device, &stage_create_info, nullptr, &shader_module) != VK_SUCCESS) {
                crash("Failed to create shader");
            }

            VkPipelineShaderStageCreateInfo pipeline_stage_create_info {};
            pipeline_stage_create_info.sType = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO;
            pipeline_stage_create_info.stage = vk_shader_stage;
            pipeline_stage_create_info.module = shader_module;
            pipeline_stage_create_info.pName = "main";

            res.stages.push_back(pipeline_stage_create_info);
        }

        return res;
    }

    void destroy_shaders(const LogicalDevice &device, const PreparedShaderSet &shaders) {
        for (const auto &shader : shaders.stages) {
            vkDestroyShaderModule(device.logical_device, shader.module, nullptr);
        }
    }
}

File: engine/modules/render_vulkan/src/renderer/shader_mgmt.rs
use std::collections::HashMap;
use glslang::{SpirvVersion, VulkanVersion};
use argus_logging::debug;
use argus_render::common::{Shader, ShaderStage};
use argus_render::constants::*;
use argus_resman::{Resource, ResourceManager};
use argus_shadertools::{compile_glsl_to_spirv, GlslCompileError, ShaderReflectionInfo};
use vk_wrapper::vk;
use crate::LOGGER;

pub(crate) struct PreparedShaderSet<'ctx> {
    pub(crate) stages: Vec<vk::PipelineShaderStageCreateInfo<'ctx>>,
    pub(crate) reflection: ShaderReflectionInfo,
}

struct ShaderCompilationResult {
    shaders: Vec<Shader<u32>>,
    reflection: ShaderReflectionInfo,
}

fn compile_glsl_shaders(shaders: &[Shader<u8>])
    -> Result<ShaderCompilationResult, GlslCompileError> {
    if shaders.is_empty() {
        return Ok(ShaderCompilationResult {
            shaders: Vec::new(),
            reflection: ShaderReflectionInfo::default(),
        });
    }

    let uids: Vec<_> = shaders.iter().map(|shader| shader.get_uid()).collect();
    let orig_shaders_map: HashMap<_, _> =
        shaders.iter().map(|shader| (shader.get_stage(), shader)).collect();
    let glsl_sources_map = &shaders
        .iter()
        .map(|shader| {
            (
                to_shadertools_stage(&shader.get_stage()),
                String::from_utf8(shader.get_source().to_vec()).unwrap(),
            )
        })
        .collect();

    debug!(LOGGER, "Compiling SPIR-V from shader set {:?}", uids);

    let compile_res = compile_glsl_to_spirv(
        glsl_sources_map,
        glslang::Target::Vulkan {
            version: VulkanVersion::Vulkan1_2,
            spirv_version: SpirvVersion::SPIRV1_5,
        },
        450,
    )?;

    let refl_info = ShaderReflectionInfo {
        attribute_locations: compile_res.inputs,
        output_locations: compile_res.outputs,
        uniform_variable_locations: compile_res.uniforms,
        buffer_locations: compile_res.buffers,
        ubo_bindings: compile_res.ubo_bindings,
        ubo_instance_names: compile_res.ubo_names,
    };

    let mut spirv_shaders: Vec<Shader<u32>> = Vec::new();
    let mut bytecode = compile_res.bytecode;
    for (stage, orig) in &orig_shaders_map {
        let spirv_src = bytecode.remove(&to_shadertools_stage(stage)).unwrap();
        let spirv_shader = Shader::new(
            orig.get_uid(),
            orig.get_stage(),
            spirv_src,
        );
        spirv_shaders.push(spirv_shader);
    }

    Ok(ShaderCompilationResult {
        shaders: spirv_shaders,
        reflection: refl_info,
    })
}

pub(crate) fn prepare_shaders<'ctx>(device: &'ctx vk::Device, shader_uids: &[impl AsRef<str>])
                                    -> Result<PreparedShaderSet<'ctx>, String> {
    let mut shader_resources: Vec<Resource> = Vec::with_capacity(shader_uids.len());
    let mut loaded_shaders: Vec<Shader<u8>> = Vec::with_capacity(shader_uids.len());
    let mut have_vert = false;
    let mut have_frag = false;
    for shader_uid in shader_uids {
        let Ok(shader_res) = ResourceManager::instance().get_resource(shader_uid) else {
            return Err(format!("Failed to load shader {}", shader_uid.as_ref()));
        };
        let Some(shader) = shader_res.get::<Shader<u8>>() else {
            return Err("Resource UID does not refer to shader".to_string());
        };

        shader_resources.push(shader_res.clone());
        loaded_shaders.push(shader.clone());

        if shader.get_stage() == ShaderStage::Vertex {
            have_vert = true;
        } else if shader.get_stage() == ShaderStage::Fragment {
            have_frag = true;
        }
    }

    if !have_vert {
        let Ok(vert_res) = ResourceManager::instance().get_resource(SHADER_STD_VERT) else {
            return Err(format!("Failed to load built-in shader {SHADER_STD_VERT}"));
        };
        let Some(vert) = vert_res.get::<Shader<u8>>() else {
            return Err(format!(
                "Resource UID '{}' does not refer to a shader",
                vert_res.get_prototype().uid,
            ));
        };
        shader_resources.push(vert_res.clone());
        loaded_shaders.push(vert.clone());
    }
    if !have_frag {
        let Ok(frag_res) = ResourceManager::instance().get_resource(SHADER_STD_FRAG) else {
            return Err(format!("Failed to load built-in shader {}", SHADER_STD_FRAG));
        };
        let Some(frag) = frag_res.get::<Shader<u8>>() else {
            return Err(format!(
                "Resource UID '{}' does not refer to a shader",
                frag_res.get_prototype().uid,
            ));
        };
        shader_resources.push(frag_res.clone());
        loaded_shaders.push(frag.clone());
    }

    //TODO: handle native SPIR-V shaders too
    let comp_res = compile_glsl_shaders(&loaded_shaders)
        .map_err(|err| format!("Failed to compile GLSL shaders {:?}", err))?;

    let mut compiled_shaders = Vec::new();

    for shader in comp_res.shaders {
        let stage = shader.get_stage();
        let spirv_src = shader.get_source();

        let vk_shader_stage = match stage {
            ShaderStage::Vertex => vk::ShaderStageFlags::VERTEX,
            ShaderStage::Fragment => vk::ShaderStageFlags::FRAGMENT,
        };

        let stage_create_info = vk::ShaderModuleCreateInfo::default()
            .code(spirv_src);

        let shader_module = vk::ShaderModule::create(device, stage_create_info)
            .map_err(|err| err.to_string())?;

        let pipeline_stage_create_info = vk::PipelineShaderStageCreateInfo::default()
            .stage(vk_shader_stage)
            .module(shader_module)
            .name(c"main");

        compiled_shaders.push(pipeline_stage_create_info);
    }

    Ok(PreparedShaderSet {
        stages: compiled_shaders,
        reflection: comp_res.reflection,
    })
}

fn to_shadertools_stage(stage: &ShaderStage) -> glslang::ShaderStage {
    match stage {
        ShaderStage::Vertex => glslang::ShaderStage::Vertex,
        ShaderStage::Fragment => glslang::ShaderStage::Fragment,
    }
}
